File: MediatorPattern/MediatorDemo.py
from __future__ import annotations


class User:
    def __init__(self, id: int, name: str) -> None:
        self.id: int = id
        self.name: str = name
        self.user_list: dict[int, User] = {}

    def add_friend(self, friend: User) -> None:
        self.user_list.update({friend.id: friend})

    def message_all(self, message: str) -> None:
        for user in self.user_list.values():
            user.send_message(f"From [{self.name}]: {message}")

    def message_one(self, id: int, message: str)  -> None:
        user = self.user_list.get(id)
        if user is not None:
            user.send_message(f"From [{self.name}] {message}")

    def send_message(self, msg: str) -> None:
        print(f"[{self.name}] got a message {msg}")


eric = User(1, "Eric Gamma")
john = User(2, "John Vlissides")
ralph = User(3, "Ralph Johnson")
richard = User(4, "Richard Helm")

eric.add_friend(john)
eric.add_friend(ralph)
eric.add_friend(richard)

john.add_friend(eric)
john.add_friend(ralph)
john.add_friend(richard)

ralph.add_friend(eric)
ralph.add_friend(john)
ralph.add_friend(richard)

richard.add_friend(eric)
richard.add_friend(john)
richard.add_friend(ralph)

eric.message_all("Hey everybody")
ralph.message_one(1, "Hi Eric")

File: MediatorPattern/MediatorDemo.java
import java.text.MessageFormat;
import java.util.HashMap;
import java.util.Map;

import javax.print.attribute.standard.Media;

class Mediator {
  Map<Integer, User> userList = new HashMap<>();

  public void addFriend(User friend) {
    userList.put(friend.id, friend);
  }

  public void messageAll(String message) {
    for (var user : userList.entrySet()) {
      // System.out.println(user.getValue().name);
      user.getValue().sendMessage(MessageFormat.format("From [{0}]: {1}", user.getValue().name, message));
    }
  }

  public void messageOne(int id, String message) {
    var user = userList.get(id);
    if (user != null)
      user.sendMessage(MessageFormat.format("From [{0}] {1}", user.name, message));
  }

}

class User {
  int id;
  String name;
  private Mediator mediator;

  public User(int id, String name, Mediator mediator) {
    this.id = id;
    this.name = name;
    this.mediator = mediator;
    this.mediator.addFriend(this);

  }

  public void messageAll(String message) {
    this.mediator.messageAll(message);

  }

  public void messageOne(int id, String message) {
    this.mediator.messageOne(id, message);
  }

  void sendMessage(String msg) {
    System.out.println(MessageFormat.format("[{0}] got a message {1}", this.name, msg));
  }

}

public class MediatorDemo {

  public static void main(String[] args) {
    var mediator = new Mediator();

    var eric = new User(1, "Eric Gamma", mediator);
    var john = new User(2, "John Vlissides", mediator);
    var ralph = new User(3, "Ralph Johnson", mediator);
    var richard = new User(4, "Richard Helm", mediator);

    // eric.addFriend(john);
    // eric.addFriend(ralph);
    // eric.addFriend(richard);

    // john.addFriend(eric);
    // john.addFriend(ralph);
    // john.addFriend(richard);

    // ralph.addFriend(eric);
    // ralph.addFriend(john);
    // ralph.addFriend(richard);

    // richard.addFriend(eric);
    // richard.addFriend(john);
    // richard.addFriend(ralph);

    eric.messageAll("Hey everybody");
    ralph.messageOne(1, "Hi Eric");
  }
}
